### PR TITLE
feat(mcp-server): Add cursor pagination and enhanced component validation

### PR DESCRIPTION
**Pull request by mahmoudmoravej**
## Overview

This PR significantly enhances the Hopper MCP server with comprehensive improvements focused on scalability, validation, and developer experience. The changes introduce robust cursor-based pagination for handling large content, comprehensive component structure validation, and enhanced documentation tools.

### 🔄 **Cursor-Based Pagination System**
- **New utility**: `cursor-pagination.ts` - Implements robust cursor-based pagination for large documentation content
- **Intelligent content chunking**: Handles large documentation files with configurable page sizes and automatic content splitting
- **Secure cursor encoding/decoding**: Base64-encoded cursors with SHA-256 content hash validation to ensure data integrity
- **Token-aware pagination**: Converts token limits to character estimates (3.5 chars/token) for optimal AI model performance
- **Pagination state management**: Maintains page size consistency throughout pagination sessions
- **Enhanced tools integration**: `get_design_tokens` and `get_guide` tools now support pagination with `page_size` and `cursor` parameters

### 🛡️ **Comprehensive Component Structure Validation**
- **Enhanced emoji detection**: Added `emoji-regex` dependency for accurate Unicode emoji detection in JSX content and attributes
- **Native HTML element validation**: Prevents usage of native HTML elements (`div`, `span`, `button`, etc.) in favor of Hopper components
- **Style prop validation**: Strong validation against `className` and `style` props with detailed guidance toward Hopper design tokens
- **Improved error reporting**: Enhanced error messages with precise line/column information and actionable suggestions
- **Performance optimizations**: Moved regex and Set creation to module scope to improve validation performance

### 🎯 **Enhanced Figma Integration**
- **New prompt**: `generate_code_from_figma_design` - Comprehensive prompt for generating Hopper components from Figma designs
- **Multi-step validation workflow**: Integrates Figma MCP tools with Hopper validation for pixel-perfect implementations
- **Design token enforcement**: Ensures all generated code uses Hopper design tokens instead of raw CSS values

### 📚 **Improved Documentation & Content Management**
- **Centralized tool descriptions**: New `tools-descriptions.ts` utility for consistent and detailed tool documentation
- **Enhanced content handling**: Better support for large documentation files with automatic pagination
- **Improved error handling**: More robust markdown file parsing with detailed error messages
- **Resource optimization**: Better content delivery with support for paginated resources

### **Architecture & Performance**
- **Modular design**: Better separation of concerns with dedicated utility files
- **Performance optimizations**: Reduced redundant operations, improved caching, and optimized content delivery
- **Memory efficiency**: Smart pagination prevents memory issues with large documentation content
- **Type safety**: Enhanced TypeScript types and interfaces for better development experience

### **Developer Experience**
- **Better error messages**: More descriptive validation errors with common issue suggestions
- **Comprehensive testing**: Extensive test coverage for emoji detection, HTML validation, and pagination features
- **Development tools**: Added pagination demonstration utilities for testing and debugging

### **Content Delivery**
- **Flexible pagination**: Supports both full content delivery and paginated responses based on request parameters
- **Content integrity**: Hash-based validation ensures paginated content consistency
- **Token optimization**: Smart token-to-character conversion for better AI model performance

### **New Files**
- `src/utils/cursor-pagination.ts` - Complete cursor-based pagination implementation
- `src/utils/tools-descriptions.ts` - Centralized tool descriptions with token count information
- `src/utils/pagination-demo.ts` - Pagination demonstration and testing utilities

### **Enhanced Files**
- `src/utils/validateComponentStructure.ts` - Major validation enhancements (emoji, HTML, style props)
- `src/utils/docs.ts` - Pagination support and improved content handling
- `src/utils/content.ts` - Enhanced content handling for arrays and pagination
- `src/utils/readMarkdownFile.ts` - Complete rewrite to support cursor pagination
- `src/tools.ts` - Updated tools with pagination parameters and improved descriptions
- `src/prompts.ts` - New Figma integration prompt
- `src/resources.ts` - Enhanced resource handling for pagination
- `src/server.ts` - Cleaned up server configuration
- `package.json` - Added `emoji-regex` dependency, version bump to 0.2.0

### **Updated Documentation & Examples**
- Multiple test files updated to use proper Hopper components instead of HTML
- Enhanced test coverage for all new validation rules
- Improved component usage examples throughout the codebase

### **Emoji Detection**
- Detects Unicode emojis in JSX content and attributes using industry-standard `emoji-regex` library
- Supports complex emoji combinations and zero-width joiners
- Provides precise location information for all detected emojis

### **Native HTML Element Prevention**
- Comprehensive validation against 20+ common HTML elements
- Detailed suggestions for Hopper component alternatives
- Helps developers maintain design system consistency

### **Style & ClassName Props**
- Strong validation against direct CSS usage
- Guidance toward Hopper design tokens and styled-system props
- Encourages proper design system patterns

## Breaking Changes

**None** - All changes are fully backward compatible. Existing functionality remains unchanged, with new features being purely additive.

## Testing & Quality Assurance

- ✅ **270+ test cases** covering all new validation rules
- ✅ **Pagination system** tested with various content sizes and edge cases  
- ✅ **Component validation** tested with complex nested structures
- ✅ **Performance testing** completed for large content handling
- ✅ **Error handling** verified for all failure scenarios
- ✅ **Server startup** and existing functionality verified unchanged

### **Cursor-Based Pagination**
```typescript
// First request - start pagination
const result1 = await server.call("get_design_tokens", {
    category: "all",
    page_size: 20000  // Start pagination with this token limit
});

// Subsequent requests - use cursor
const result2 = await server.call("get_design_tokens", {
    category: "all",
    cursor: result1.nextCursor  // Continue from where we left off
});
```

### **Enhanced Validation**
The validation now catches comprehensive issues:
- ✅ Emoji usage: `<Button>Submit 🚀</Button>` → Error with guidance
- ✅ Native HTML: `<div>Content</div>` → Error suggesting `<Div>Content</Div>`
- ✅ Style props: `<Button className="btn">Click</Button>` → Error directing to design tokens
- ✅ Complex validations: Nested components, multiple instances, and edge cases

### **Figma Integration**
```typescript
// Use the new prompt for Figma-to-Hopper conversion
const result = await server.call("generate_code_from_figma_design", {
    figma_design_url: "https://figma.com/design/..."
});
// Automatically generates Hopper-compliant code with full validation
```

## Impact & Benefits

1. **🚀 Scalability**: Handles large documentation without memory or performance issues
2. **🎯 Quality**: Ensures all generated code follows Hopper design system best practices  
3. **⚡ Performance**: Optimized content delivery and validation processes
4. **🛠️ Developer Experience**: Better error messages and guidance for faster development
5. **📖 Documentation**: Improved access to comprehensive design system information
6. **🔄 Future-Ready**: Scalable architecture supports growing documentation and features

This release transforms the Hopper MCP server into a production-ready tool capable of handling enterprise-scale documentation while ensuring the highest quality of generated code through comprehensive validation and guidance.

### CODE REVIEW

File: apps/mcp-server/src/utils/pagination-demo.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
+import { paginateContent } from "./cursor-pagination.js";
+
+// Example usage of the new cursor-based pagination
+function demonstratePagination() {
+    const longContent = "This is a very long piece of content that will be split into multiple pages. ".repeat(100);
```
[the remark this answers is not in the document]
Comment (mahmoudmoravej) [reply apps/mcp-server/src/utils/pagination-demo.ts:5]: Left my mistake. AI Genereated stuff. Removed.
